### PR TITLE
Micromanagement arm movement

### PR DESCRIPTION
**Pull request by MatthijsHofele**

### CODE REVIEW

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/PlayerController.cs [hunk 1]
```diff
@@ -37,6 +37,8 @@ public class PlayerController : MonoBehaviour
     [SerializeField] public Transform AnimatedChest;
     [SerializeField] public Rigidbody PhysicalTorso;
     [SerializeField] public Rigidbody PhysicalHead;
+    [SerializeField] public GameObject HandL;
+    [SerializeField] public GameObject HandR;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/PlayerController.cs:41]:
```suggestion
    public GameObject HandLeft;
    public GameObject HandRight;
```

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.InputSystem;
+using Cinemachine;
+
+public class GrabModule : MonoBehaviour
+{
+	private bool hasJoint;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs:9]:
```suggestion
    private bool hasJoint;
```

File: WobblyStealthGame/Assets/HandIk.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+using UnityEngine;
+
+public class HandIk : MonoBehaviour
+{
+    protected Animator animator;
+    
+    public static bool ikActiveLeft = false;
+    public static bool ikActiveRight = false;
+    public Transform rightHandObj = null;
+    public Transform leftHandObj = null;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:12]: These two properties aren't used outside of this class. Shouldn't they be private? if not they require to follow PascalCase.

File: WobblyStealthGame/Assets/HandIk.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class HandIk : MonoBehaviour
+{
+    protected Animator animator;
+    
+    public static bool ikActiveLeft = false;
+    public static bool ikActiveRight = false;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:10]:
```suggestion
    public static bool IkActiveLeft = false;
    public static bool IkActiveRight = false;
```

File: WobblyStealthGame/Assets/HandIk.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    void Start () 
+    {
+        animator = GetComponent<Animator>();
+    }
+
+    void OnAnimatorIK()
+    {
+        if (animator) 
+        {   
+            if (ikActiveLeft)
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:23]:
```suggestion
            if (IkActiveLeft) 
```

File: WobblyStealthGame/Assets/HandIk.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                if (leftHandObj != null) 
+                {
+                    animator.SetIKPositionWeight(AvatarIKGoal.LeftHand, 1);
+                    animator.SetIKRotationWeight(AvatarIKGoal.LeftHand, 1);  
+                    animator.SetIKPosition(AvatarIKGoal.LeftHand,leftHandObj.position);
+                    animator.SetIKRotation(AvatarIKGoal.LeftHand,leftHandObj.rotation);
+                }        
+            }
+
+            if (ikActiveRight)
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:35]:
```suggestion
            if (IkActiveRight)
```

File: WobblyStealthGame/Assets/HandIk.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        {   
+            if (ikActiveLeft) 
+            {
+
+                if (leftHandObj != null) 
+                {
+                    animator.SetIKPositionWeight(AvatarIKGoal.LeftHand, 1);
+                    animator.SetIKRotationWeight(AvatarIKGoal.LeftHand, 1);  
+                    animator.SetIKPosition(AvatarIKGoal.LeftHand,leftHandObj.position);
+                    animator.SetIKRotation(AvatarIKGoal.LeftHand,leftHandObj.rotation);
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:31]:
```suggestion
                    animator.SetIKPosition(AvatarIKGoal.LeftHand, leftHandObj.position);
                    animator.SetIKRotation(AvatarIKGoal.LeftHand, leftHandObj.rotation);
```

File: WobblyStealthGame/Assets/HandIk.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            if (ikActiveRight)
+            {
+                if (rightHandObj != null) 
+                {
+                    animator.SetIKPositionWeight(AvatarIKGoal.RightHand, 1);
+                    animator.SetIKRotationWeight(AvatarIKGoal.RightHand, 1);  
+                    animator.SetIKPosition(AvatarIKGoal.RightHand,rightHandObj.position);
+                    animator.SetIKRotation(AvatarIKGoal.RightHand,rightHandObj.rotation);
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/HandIk.cs:42]:
```suggestion
                    animator.SetIKPosition(AvatarIKGoal.RightHand, rightHandObj.position);
                    animator.SetIKRotation(AvatarIKGoal.RightHand, rightHandObj.rotation);
```

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/InputModule.cs [hunk 9]
```diff
@@ -57,6 +57,28 @@ public void OnToggleBalance()
         }
     }
 
+    public delegate void onGrabLeftDelegate();
+    public onGrabLeftDelegate OnGrabLeftDelegates { get; set; }
+    public void OnGrabLeft()
+    {
+        HandIk.ikActiveLeft = !HandIk.ikActiveLeft;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/InputModule.cs:64]:
```suggestion
        HandIk.IkActiveLeft = !HandIk.IkActiveLeft;
```

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/InputModule.cs [hunk 10]
```diff
@@ -57,6 +57,28 @@ public void OnToggleBalance()
...
+        {
+            OnGrabLeftDelegates?.Invoke();
+        }
+    }
+
+    public delegate void onGrabRightDelegate();
+    public onGrabRightDelegate OnGrabRightDelegates { get; set; }
+    public void OnGrabRight()
+    {
+        HandIk.ikActiveRight = !HandIk.ikActiveRight;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/InputModule.cs:75]:
```suggestion
        HandIk.IkActiveRight = !HandIk.IkActiveRight;
```

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    private bool grabbing;
+    private GameObject holdingItem;
+
+    void OnCollisionEnter(Collision collision)
+    {
+    	if (!grabbing || hasJoint || collision.gameObject.name == "CameraCollider" || (!collision.gameObject.GetComponent<BoxCollider>() && !collision.gameObject.GetComponent<CapsuleCollider>()) && !collision.gameObject.GetComponent<MeshCollider>().convex) return;
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs:15]:
```suggestion
    	if (!grabbing || hasJoint || collision.gameObject.name == "CameraCollider" || (!collision.gameObject.GetComponent<BoxCollider>() && !collision.gameObject.GetComponent<CapsuleCollider>()) && !collision.gameObject.GetComponent<MeshCollider>().convex) 
    	{
            return;
        }
```

File: WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        holdingItem = collision.gameObject;
+    }
+
+    public void Grab()
+    {
+        grabbing = !grabbing;
+        if (!grabbing)
+        {
+            hasJoint = false;
+            if (holdingItem && holdingItem.GetComponent<FixedJoint>()) Destroy(holdingItem.GetComponent<FixedJoint>());
```
Comment (ahoefnagel) [inline WobblyStealthGame/Assets/Scripts/Input/ActiveRagdoll/GrabModule.cs:30]:
```suggestion
            if (holdingItem && holdingItem.GetComponent<FixedJoint>()) 
            {
                Destroy(holdingItem.GetComponent<FixedJoint>());
            }
```

**Review by ahoefnagel — CHANGES_REQUESTED** [review]

**Review by ahoefnagel — APPROVED** [review]

**Review by ahoefnagel — APPROVED** [review]

**Review by SjonsDePlons — APPROVED** [review]

**Pull request merged**